modules/auxiliary/gather/shodan_search.rb

```diff
@@ -117,92 +119,113 @@ def run
+            next
+          end
+          results[page] = page_result
+          page += 1
+        end
+      end
+      # Save the results to this table
+      tbl = Rex::Text::Table.new(
+        'Header'  => 'Search Results',
+        'Indent'  => 1,
+        'Columns' => ['IP:Port', 'City', 'Country', 'Hostname']
+      )
 
-    # Organize results and put them into the table and database
-    regex = datastore['REGEX'] if datastore['REGEX']
-    results.each do |page|
-      page['matches'].each do |host|
-        city = host['location']['city'] || 'N/A'
-        ip   = host['ip_str'] || 'N/A'
-        port = host['port'] || ''
-        country = host['location']['country_name'] || 'N/A'
-        hostname = host['hostnames'][0]
-        data = host['data']
-
-        report_host(:host     => ip,
-                    :name     => hostname,
-                    :comments => 'Added from Shodan',
-                    :info     => host['info']
-                    ) if datastore['DATABASE']
-
-        report_service(:host => ip,
-                       :port => port,
-                       :info => 'Added from Shodan'
-                       ) if datastore['DATABASE']
-
-        if ip =~ regex ||
-          city =~ regex ||
-          country =~ regex ||
-          hostname =~ regex ||
-          data =~ regex
-          # Unfortunately we cannot display the banner properly,
-          # because it messes with our output format
-          tbl << ["#{ip}:#{port}", city, country, hostname]
+      # Organize results and put them into the table and database
+      regex = datastore['REGEX'] if datastore['REGEX']
+      results.each do |page|
+        page['matches'].each do |host|
+          city = host.dig('location','city') || 'N/A'
+          ip   = host.fetch('ip_str', 'N/A')
+          port = host.fetch('port', '')
+          country = host.dig('location','country_name') || 'N/A'
+          hostname = host.dig('hostnames',0)
+          data = host.dig('data')
+
+          report_host(:host     => ip,
+                      :name     => hostname,
+                      :comments => 'Added from Shodan',
+                      :info     => host.dig('info')
+                      ) if datastore['DATABASE']
+
+          report_service(:host => ip,
+                        :port => port,
+                        :info => 'Added from Shodan'
+                        ) if datastore['DATABASE']
+
+          if ip =~ regex ||
+            city =~ regex ||
+            country =~ regex ||
+            hostname =~ regex ||
+            data =~ regex
+            # Unfortunately we cannot display the banner properly,
+            # because it messes with our output format
+            tbl << ["#{ip}:#{port}", city, country, hostname]
+          end
         end
       end
+      #Show data and maybe save it if needed
+      print_line()
+      print_line("#{tbl}")
+      save_output(tbl) if datastore['OUTFILE']
+    end
+    if datastore['FACETS']
```

Comment:
Can we use previously defined variable `facets`? 
```suggestion
    if facets
```

Reply:
It seems like facets use the same parameters than the filters, I noted it in the README.